### PR TITLE
barebones transaction work 

### PR DESCRIPTION
**Pull request by relaxolotl**
mostly a scratchpad of stuff being done to support bare bones transactions, plan is to break this up into smaller PRs

### CODE REVIEW

**Comment by relaxolotl** [comment]
oh no

File: include/sentry.h [hunk 1]
```diff
@@ -1226,6 +1226,50 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
...
+/**
+ * Constructs a new transaction context to be passed
+ * into `sentry_start_transaction`.
+ */
+SENTRY_EXPERIMENTAL_API sentry_value_t sentry_value_new_transaction_context(
+    const char *name);
+
+/**
+ * Sets the name of a transaction on a `transaction_context`.
+ * Returns 1 if the name provided is invalid (e.g. empty or nil).
```
Comment (Swatinem) [inline include/sentry.h:1240]: Is this consistent with other similar APIs? I don’t quite remember if we even validate invalid values?

File: src/sentry_core.c [hunk 2]
```diff
@@ -633,3 +656,94 @@ sentry_set_level(sentry_level_t level)
...
+sentry_value_t
+sentry_start_transaction(sentry_value_t tx_cxt)
+{
+    // TODO: stuff transaction into the scope
+    bool should_sample = sentry__should_sample_transaction(tx_cxt);
+
+    sentry_value_t tx = sentry_value_new_event();
+
+    // TODO: does it matter that this is sampled on creation vs on send? can we
+    // just sample on send like other events?
```
Comment (Swatinem) [inline src/sentry_core.c:686]: good Q. I think since the sampling decision can also be set from the outside, it should be defined here?

**Review by Swatinem — COMMENTED** [review]
this looks very nice so far!

File: src/sentry_core.c [hunk 3]
```diff
@@ -633,3 +656,94 @@ sentry_set_level(sentry_level_t level)
...
+
+    sentry_value_decref(tx_cxt);
+
+    return tx;
+}
+
+void
+sentry_transaction_finish(sentry_value_t tx)
+{
+    sentry_value_set_by_key(tx, "type", sentry_value_new_string("transaction"));
```
Comment (loewenheim) [inline src/sentry_core.c:726]: Why are we setting the type here?

**Comment by loewenheim** [comment]
I don't quite understand how the more strongly typed API of `sentry_tracing_stub.h` is meant to interact with the JSONly typed API elsewhere.

File: include/sentry.h (hunk 1 above)
Comment (relaxolotl) [reply include/sentry.h:1240]: oh yeah, good catch. [python](https://github.com/getsentry/sentry-python/blob/dd0efc08414ee2ef1a5f22d2cc4e243b54a1b455/sentry_sdk/tracing.py#L565-L568) and [js](https://github.com/getsentry/sentry-javascript/blob/5bba9b44b47ffef3c199ed6361df046009b789c4/packages/tracing/src/transaction.ts#L95-L96) merely warn and set the value to `<unlabeled transaction>`, so i'll fall back to that.

Comment (Swatinem) [reply include/sentry.h:1240]: When I was experimenting with this in Rust, the server (relay?) would also default this. So the SDK is free to leave off the name completely (or make it empty) instead of defaulting it.

File: src/sentry_core.c (hunk 3 above)
Comment (relaxolotl) [reply src/sentry_core.c:726]: good question; if you're wondering why it's set in finish of all places, i'd have to say that there's no real rhyme or reason behind it besides the idea that the transactions _shouldn't_ be treated like a regular event (ie require checking of the type field) elsewhere in the code after their creation. that's likely a naive assumption given the lack of typing we have for transactions, though. now that i've written this out, i might just set this field during initialization to be safe.

if you're wondering why this field is getting set at all (and why it isn't being set for events), it's the discriminant in "events" (https://develop.sentry.dev/sdk/event-payloads/types/#eventtype). tl;dr SDKs send two types of events, error monitoring-related events and transactions. the former is the default and doesn't need an explicit `type`, the latter must have the `type` field set to transaction for it to be treated as such.

**Comment by relaxolotl** [comment]
> I don't quite understand how the more strongly typed API of sentry_tracing_stub.h is meant to interact with the JSONly typed API elsewhere.

apologies for the confusion; it was probably a mistake for me to commit this to begin with. it was probably really confusing to see these types and then discover that none of them are used anywhere in the code. i unfortunately haven't spent much time thinking about how i could integrate these types internally into the implementation while keeping the JSON-ly typed API that's currently exposed to users. 

my mulling on this has terminated at "golly it would be a massive pain and waste of CPU cycles to be constantly converting between a `sentry_value_t` and a `sentry_transaction_or_other_t`". if you (or anybody else) have any ideas on how to tackle this challenge i'm all ears.

File: examples/example.c [hunk 4]
```diff
@@ -208,6 +211,20 @@ main(int argc, char **argv)
         sentry_capture_event(event);
     }
 
+    if (has_arg(argc, argv, "capture-transaction")) {
+        sentry_value_t tx_ctx
+            = sentry_value_new_transaction_context("I'm a little teapot");
+        sentry_transaction_context_set_operation(
+            tx_ctx, "Short and stout here is my handle and here is my spout");
```
Comment (flub) [inline examples/example.c:218]: More of an api question, but i wonder if it makes sense to instead to always add this as a second argument to `sentry_value_new_transaction_context("op", "name")`, this would be similar to python's `start_transaction(op, name)`. it was my understanding that you really want to have `op` filled in, if we do not want to make them obligatory we can still make them accept `NULL` for them.

also happy to discuss this on the notion page if you prefer

File: include/sentry.h [hunk 5]
```diff
@@ -1226,6 +1226,51 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
 SENTRY_EXPERIMENTAL_API double sentry_options_get_traces_sample_rate(
     sentry_options_t *opts);
 
+/* -- Performance Monitoring/Tracing APIs -- */
+
+/**
+ * Constructs a new transaction context to be passed
+ * into `sentry_start_transaction`.
```
Comment (flub) [inline include/sentry.h:1233]: could we describe what the name parameter is?  something like
```
*name* is the transaction name, this describes the transaction and can also be used
to search for the transaction in sentry.
```

File: include/sentry.h [hunk 6]
```diff
@@ -1226,6 +1226,51 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
...
+ */
+SENTRY_EXPERIMENTAL_API sentry_value_t sentry_value_new_transaction_context(
+    const char *name);
+
+/**
+ * Sets the name of a transaction on a `transaction_context`.
```
Comment (flub) [inline include/sentry.h:1239]: same, i'd like to describe what name is for

File: include/sentry.h [hunk 7]
```diff
@@ -1226,6 +1226,51 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
...
+ */
+SENTRY_EXPERIMENTAL_API void sentry_transaction_context_set_name(
+    sentry_value_t transaction_context, const char *name);
+
+/**
+ * Sets the operation of a transaction on a `transaction_context`.
```
Comment (flub) [inline include/sentry.h:1245]: similar, maybe describe what operation is?

File: include/sentry.h [hunk 8]
```diff
@@ -1226,6 +1226,51 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
...
+ */
+SENTRY_EXPERIMENTAL_API void sentry_transaction_context_set_operation(
+    sentry_value_t transaction_context, const char *operation);
+
+/**
+ * Sets the sampled field a transaction on a `transaction_context`.
+ * When turned on, the transaction will bypass all sampling
+ * options and always be sent to sentry. If this is explicitly
+ * turned off in the `transaction_context`, the transaction will
+ * never be sent to sentry.
```
Comment (flub) [inline include/sentry.h:1255]: because `sampled` is such a confusing word, could we explicitly describe: `1: is sent to sentry; 0: is **not** sent to sentry`?

File: include/sentry.h [hunk 9]
```diff
@@ -1226,6 +1226,51 @@ SENTRY_EXPERIMENTAL_API void sentry_options_set_traces_sample_rate(
...
+/**
+ * Starts a new transaction based on the provided context, restored
+ * from an external integration (i.e. a span from a different SDK)
+ * or manually constructed by a user.
+ */
+SENTRY_EXPERIMENTAL_API sentry_value_t sentry_start_transaction(
+    sentry_value_t transaction_context);
+
+/**
+ * Finishes a transaction.
```
Comment (flub) [inline include/sentry.h:1269]: is it now put on the transport to be sent to sentry?  I'd add this in the description.

File: src/sentry_core.c (hunk 2 above)
Comment (flub) [reply src/sentry_core.c:693]: potentially adding spans could become a no-op if sampled on creation?  saves some cpu

File: src/sentry_value.c [hunk 10]
```diff
@@ -1124,6 +1124,44 @@ sentry_value_new_stacktrace(void **ips, size_t len)
...
+
+    return transaction_context;
+}
+
+void
+sentry_transaction_context_set_name(
+    sentry_value_t transaction_context, const char *name)
+{
+    sentry_value_t sv_name = sentry_value_new_string(name);
+    if (sentry__string_eq(name, "") || sentry_value_is_null(sv_name)) {
```
Comment (flub) [inline src/sentry_value.c:1142]: should the order around `||` not be inversed?  or is `sentry__string_eq` smart enough?

File: tests/unit/test_tracing.c [hunk 11]
```diff
@@ -33,3 +34,32 @@ SENTRY_TEST(basic_tracing_context)
     sentry_value_decref(trace_context);
     sentry_value_decref(span);
 }
+
+SENTRY_TEST(basic_transaction_context)
+{
+    sentry_value_t tx_cxt = sentry_value_new_transaction_context("");
```
Comment (flub) [inline tests/unit/test_tracing.c:40]: do you also want to test `NULL`?

File: examples/example.c (hunk 4 above)
Comment (relaxolotl) [reply examples/example.c:218]: i don't have any particularly strong feelings on the matter - if it's more convenient to require an `op` and it lines up with the sort of behaviour we would like to encourage from users, then this can definitely be done. the change will most likely be made in https://github.com/getsentry/sentry-native/pull/619. 

i would like to call out the description on this PR, which is still applicable to the contents of its changes: the main intent of this PR is to simply expose the work i'm doing to the rest of the team, but none of it is really "PR-ready" from my POV. i can address the comments made on this diff and any discussion on parts of this that aren't already opened up in existing PRs are welcome, but given the genuine WIP nature of the changes on here some things are expected to be fairly rough (e.g. the documentation).

Comment (relaxolotl) [reply examples/example.c:218]: probably on me for making it possible to even see those parts of the diffs though...

File: src/sentry_value.c (hunk 10 above)
Comment (relaxolotl) [reply src/sentry_value.c]: ah yeah, these probably need to be reversed.

File: include/sentry.h (hunk 9 above)
Comment (relaxolotl) [reply include/sentry.h:1276]: perhaps? i'm not sure about the details myself; this doesn't really do much of anything different from `capture_event` whose description is just `Sends a sentry event.`

File: include/sentry.h (hunk 5 above)
Comment (relaxolotl) [reply include/sentry.h]: i'm a little reluctant to provide too much product guidance here as existing documentation tends to defer to documentation on docs.sentry.io on that front (ie `sentry_value_new_event`, `sentry_value_new_message_event`, `sentry_value_new_exception`, etc), preferring to mostly explain implementation-specific details. i'll admit that we can't really do the same for performance monitoring as its documentation isn't compiled in one easily referenceable location in the same way other types of messages are, making it difficult for us to do the same.

i've pushed an attempt to reach some sort of compromise between a full explanation and the approach used by existing documentation in #619.

**Comment by relaxolotl** [comment]
got started in creating some transaction-equivalents to envelope helpers that can also be found in other SDKs. the term "event" is starting to gain two different definitions though: the blanket "Event" that covers all types of items that may be stuffed into an envelope, and the "Event" that is a specific variant of the blanket "Event" which is one of those very items that can be placed in an envelope, and is a sibling to a Transaction.

File: include/sentry.h (hunk 9 above)
Comment (flub) [reply include/sentry.h:1284]: so `Finishes a transaction and sends it`.?

File: examples/example.c (hunk 4 above)
Comment (flub) [reply examples/example.c:218]: i may be commenting on the wrong PRs.  apologies, i've not been paying enough attention on what the various PRs are for

**Comment by relaxolotl** [comment]
now outdated; other PRs are now ahead of this work and i've switched to a different (local) scratchpad branch.
